### PR TITLE
FIX noqa by using register_reader(), address many side effects

### PR DESCRIPTION
**Pull request by jalew188**
Review this PR after #171 

Changes:

- Remove `alphabase.io.psm_reader` as all functionalities are already in `alphabase.psm_reader` module.
- register_reader() for all reader modules
- register_decoy() for protein_level_decoy

### CODE REVIEW

**Review by mschwoer — APPROVED** [review]
LGTM.. reader registration is much cleaner now

File: alphabase/psm_reader/__init__.py [hunk 1]
```diff
@@ -1,23 +1,58 @@
-# ruff: noqa: F401 TODO remove import side effects (i.e. change way readers are registered)
-# to register all readers into .psm_reader.psm_reader_provider
+__all__ = [
```
[the remark this answers is not in the document]
Comment (jalew188) [reply alphabase/psm_reader/__init__.py:1]: The point is, we can import readers directly from `alphabase.psm_reader` thus users do not need to remember the sub-module names, e.g. alphabase.psm_reader.`dia_psm_reader`, `maxquant_reader` and so on. But adding readers in `__init__.py` will cause noqa F401 issues, so `pre-commit` suggests using `__all__`.

File: alphabase/peptide/fragment.py [hunk 2]
```diff
@@ -1321,7 +1321,9 @@ def _calc_fragment_cardinality(
         fragment_mz,
         fragment_cardinality,
     ):
-        elution_group_idx[0]  # noqa TODO check for potential bug
+        if len(elution_group_idx) == 0:
+            return
+        elution_group_idx[0]
```
[the remark this answers is not in the document]
Comment (GeorgWa) [reply alphabase/peptide/fragment.py:1326]: That's indeed weird. I wrote this, will fix it 👍

[the remark this answers is not in the document]
Comment (GeorgWa) [reply alphabase/peptide/fragment.py:1326]: That's indeed weird. I wrote this, will fix it 👍

File: alphabase/spectral_library/base.py [hunk 3]
```diff
@@ -309,7 +309,9 @@ def append_decoy_sequence(self):
         from alphabase.spectral_library.decoy import decoy_lib_provider
```
[the remark this answers is not in the document]
Comment (jalew188) [reply alphabase/spectral_library/base.py:309]: No, there will be a cyclic imports...

File: alphabase/peptide/fragment.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jalew188) [reply alphabase/peptide/fragment.py:1326]: Is checking `len==0` enough?

File: alphabase/psm_reader/__init__.py [hunk 4]
```diff
@@ -1,23 +1,58 @@
...
-from alphabase.psm_reader.maxquant_reader import MaxQuantReader
-from alphabase.psm_reader.pfind_reader import pFindReader
-from alphabase.psm_reader.msfragger_reader import MSFragger_PSM_TSV_Reader
-from alphabase.psm_reader.sage_reader import SageReaderTSV, SageReaderParquet
 
-try:
-    from alphabase.psm_reader.msfragger_reader import MSFraggerPepXML  # noqa: F401 TODO remove import side effects (registering the reader)
-except ImportError:
-    pass
+register_ap_readers()
```
[the remark this answers is not in the document]
Comment (jalew188) [reply alphabase/psm_reader/__init__.py:53]: Sure

File: alphabase/psm_reader/__init__.py [hunk 5]
```diff
@@ -1,23 +1,58 @@
...
+register_dia_readers()
+register_fragger_readers()
```
[the remark this answers is not in the document]
Comment (jalew188) [reply alphabase/psm_reader/__init__.py:55]: Thx, this is what I am looking for. Mine is still C/C++ style. We should create an issue with is for all provider-based factory classes.

[the remark this answers is not in the document]
Comment (jalew188) [reply alphabase/psm_reader/__init__.py:55]: Issue created!

File: nbs_tests/psm_reader/sage_reader.ipynb [hunk 6]
```diff
@@ -26,7 +26,8 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "from alphabase.psm_reader.sage_reader import *"
+    "from alphabase.psm_reader.sage_reader import *\n",
```
[the remark this answers is not in the document]
Comment (jalew188) [reply nbs_tests/psm_reader/sage_reader.ipynb:29]: This is in my plan when we move ipynb tests to real pytest tests

**Pull request merged**